Make remove identity equiv multithreaded

mtreinish opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

The RemoveIdentityEquiv transpiler pass analyzes each gate in the circuit to check if it's equivalent to an identity operation to check for removal. This checking was previously always performed serially but there is no data dependency between the checks so they can be performed in parallel using multithreading. While in practice the node removal can end up dominating the runtime of the pass and must be performed serially so this limits the potential speedup in cases with a lot of nodes to remove. But in practice there aren't typically many gates that get removed so there are real performance benefits available from doing the analysis in parallel.

### Details and comments

mtreinish committed:
Make remove identity equiv multithreaded

The RemoveIdentityEquiv transpiler pass analyzes each gate in the
circuit to check if it's equivalent to an identity operation to check
for removal. This checking was previously always performed serially but
there is no data dependency between the checks so they can be performed
in parallel using multithreading. While in practice the node removal
can end up dominating the runtime of the pass and must be performed
serially so this limits the potential speedup in cases with a lot of
nodes to remove. But in practice there aren't typically many gates that
get removed so there are real performance benefits available from doing
the analysis in parallel.

mtreinish commented:
To find the switchover point in the code between serial and parallel I wrote a small scale test to look at different ratios of identity gates and different sizes:
```python
import csv
import statistics
import time

import numpy as np

from qiskit.circuit import QuantumCircuit
from qiskit.converters import circuit_to_dag
from qiskit.circuit.quantumcircuitdata import CircuitInstruction
from qiskit.circuit.library import XGate, RZGate, RGate
from qiskit.transpiler.passes import RemoveIdentityEquivalent

opt_pass = RemoveIdentityEquivalent()

with open("times_parallel.csv", "w") as fd:
    writer = csv.writer(fd)
    writer.writerow(["circuit", "size", "runtime"])
    for size in np.logspace(1, 7, dtype=int):
        size = int(size)
        c_no_ident = QuantumCircuit(1)
        qubits = [c_no_ident.qubits[0]]
        for _ in range(size):
            c_no_ident._append(CircuitInstruction.from_standard(XGate._standard_gate, qubits, []))
        c_half_ident = QuantumCircuit(1)
        qubits = [c_half_ident.qubits[0]]
        for _ in range(int(size / 2)):
            c_half_ident._append(CircuitInstruction.from_standard(XGate._standard_gate, qubits, []))
        for _ in range(int(size / 2)):
            c_half_ident._append(CircuitInstruction.from_standard(RZGate._standard_gate, qubits, [0.]))
        c_full_ident = QuantumCircuit(1)
        qubits = [c_full_ident.qubits[0]]
        for _ in range(size):
            c_full_ident._append(CircuitInstruction.from_standard(RGate._standard_gate, qubits, [0., 0.]))

        d_no_ident = circuit_to_dag(c_no_ident)
        d_half_ident = circuit_to_dag(c_half_ident)
        d_full_ident = circuit_to_dag(c_full_ident)
        times = []
        for _ in range(10):
            start = time.perf_counter()
            opt_pass.run(d_no_ident)
            stop = time.perf_counter()
            times.append(stop - start)
        writer.writerow(["None", size, statistics.geometric_mean(times)])
        print(f"None size {size}: {statistics.geometric_mean(times)}")
        times = []
        for _ in range(10):
            start = time.perf_counter()
            opt_pass.run(d_half_ident)
            stop = time.perf_counter()
            times.append(stop - start)
        writer.writerow(["Half", size, statistics.geometric_mean(times)])
        print(f"Half size {size}: {statistics.geometric_mean(times)}")
        times = []
        for _ in range(10):
            start = time.perf_counter()
            opt_pass.run(d_full_ident)
            stop = time.perf_counter()
            times.append(stop - start)
        writer.writerow(["Full", size, statistics.geometric_mean(times)])
        print(f"Full size {size}: {statistics.geometric_mean(times)}")
        fd.flush()
```

Which yielded:
![equiv_plot_full](https://github.com/user-attachments/assets/ce181024-44fb-4bee-9533-dea95ce7adba)
![equiv_plot_half](https://github.com/user-attachments/assets/ef6d720b-3e34-4319-b4ab-7186ed1d86d1)
![equiv_plot_none](https://github.com/user-attachments/assets/c3b2e6c2-0239-436d-9d22-d3c904f71f89)

and I selected it based on the none case because there are rarely a lot of gates to remove as part of this pass.

mtreinish committed:
Merge remote-tracking branch 'origin/main' into parallel-identity-removal

mtreinish committed:
Avoid an allocation for node indices in parallel path

mtreinish commented:
I've run updated benchmarks on the current state of the PR to compare it to main:

<img width="5760" height="4320" alt="full" src="https://github.com/user-attachments/assets/56b8f2b0-ece3-4f4e-ab47-363d1b6af78d" />
<img width="5760" height="4320" alt="half" src="https://github.com/user-attachments/assets/2336cfb2-0045-462d-b37d-20ba23c5bb0a" />
<img width="5760" height="4320" alt="none" src="https://github.com/user-attachments/assets/fdae54e0-114e-493b-b605-9d65a23167e6" />

I'm running these on my laptop (on battery power) so small spikes and similar can probably be ignored.
These plots match my expectations the pass is basically identical in the serial regime maybe marginally slower, but not significant. But for larger circuits it is noticeably faster when parallelism. The exception is the full identity gates case which is dominated by the removal handling (and the collection into a vec of all nodes).

mtreinish committed:
Merge remote-tracking branch 'origin/main' into parallel-identity-removal

alexanderivrii commented on crates/transpiler/src/passes/remove_identity_equiv.rs:
```diff
@@ -264,25 +275,96 @@ pub fn run_remove_identity_equiv(
-        dag.remove_op_node(node);
-    }
+        let view = inst.op.view();
+        match view {
```
Hi Matt, we have done a few changes to this file since you have originally written this PR, in particular there is a function `is_identity_equiv` with similar functionality to the code here. Is it possible to combine the two, avoiding code duplication?

mtreinish replied:
Sure, I can move to use that instead. I just did a naive rebase when I updated it last night. But if there is a better function now I'll try to use that.

mtreinish replied:
Updated in: https://github.com/Qiskit/qiskit/pull/14719/commits/e156103b497e6fdff3702332561f15acd7f94c7f

mtreinish committed:
Merge remote-tracking branch 'origin/main' into parallel-identity-removal

mtreinish committed:
Use newer pass code after rebase

When this pass was updated to use main it ignored all the structural
changes in the file and was duplicating the logic of the pass which had
already been factored out into a standalone function. This commit
updates the new parallel pass to use this function and removes the
duplication. As a side-effect because the new function returns a
PyResult because of PauliEvolutionGate handling this PyResult needs to
be propogated out from the function now. This is at least until #15450
is implemented avoid the need to use python at all here.

mtreinish committed:
Add docs about threading to C API docs too

alexanderivrii commented on crates/transpiler/src/passes/remove_identity_equiv.rs:
```diff
@@ -264,25 +275,38 @@ pub fn run_remove_identity_equiv(
-    for node in remove_list {
-        dag.remove_op_node(node);
-    }
+    let process_node = |op_node: NodeIndex, inst: &PackedInstruction| {
```
nice, I didn't know that one can skip specifying the output of the closure, provided that the compiler can figure this out

alexanderivrii commented on crates/transpiler/src/passes/remove_identity_equiv.rs:
```diff
@@ -264,25 +275,38 @@ pub fn run_remove_identity_equiv(
+    };
+    let run_in_parallel = getenv_use_multiple_threads();
+    let remove_list: Vec<(NodeIndex, f64)> = if dag.num_ops() >= 5e4 as usize && run_in_parallel {
+        (0..dag.dag().node_bound())
```
I see, `5e4` is the crossover point from the PR description. Can we expose this constant at the top of the file, so that we can remember that this is a tunable parameter?

mtreinish replied:
Sure, I'll switch this to a constant and add a comment about it.

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/14719/commits/4205611d0caa8eca0d83510e20f6b8a5c593a1e2

alexanderivrii reviewed (commented):
This is a cool PR. My only question is whether you want to add a release note?

Off-topic, I was thinking about your point that node removal still remains sequential, and a quick google search shows that there are in fact graph algorithms for removing sets of nodes in parallel (though these typically require additional computations and it's not clear that the benefit outweighs the cost). On the other hand, if too many nodes need to be removed, it might be cheaper to recreate the DAG from scratch?

mtreinish committed:
Merge remote-tracking branch 'origin/main' into parallel-identity-removal

mtreinish committed:
Use a constant for PARALLEL_THRESHOLD

mtreinish committed:
Add release note

alexanderivrii reviewed (approved):
LGTM!
